src/streamlink/plugins/filmon.py

```diff
@@ -220,14 +231,28 @@ def _get_streams(self):
         vod_id = self.match.group("vod_id")
         is_group = self.match.group("is_group")
 
+        # get cookies
+        self.session.http.get(self.url)
+
         if vod_id:
+            mp4_warning_displayed = False
             data = self.api.vod(vod_id)
             for _, stream in data["streams"].items():
-                streams = HLSStream.parse_variant_playlist(self.session, stream["url"])
-                if not streams:
-                    yield stream["quality"], HLSStream(self.session, stream["url"])
+                if stream["url"].endswith(".m3u8"):
+                    streams = HLSStream.parse_variant_playlist(self.session, stream["url"])
+                    if not streams:
+                        yield stream["quality"], HLSStream(self.session, stream["url"])
+                    else:
+                        yield from streams.items()
+                elif stream["url"].endswith(".mp4"):
+                    if not mp4_warning_displayed:
+                        log.warning("Detected .mp4 video, this may not stream directly to your player. "
+                                    "In that case, try using -o/--output to save it to a file instead.")
+                        mp4_warning_displayed = True
```

Comment:
> Note: the warning for .mp4 video sources is due to the moov atom being located
> at the end of file. JW Player appears to get around this via range headers,
> sourcing data from the end of file first and then starting from the beginning
> to play the content as usual.

Are `HTTPStream`s actually needed? If the content is always available via HLS and the HTTPStream is just used by the site as a fallback, then it should be removed from the plugin. Otherwise the HTTPStream class should be subclassed and updated with the range header requests, if it's possible. Just adding a warning message isn't particularly great and should only be added as a last resort.

Reply:
Unfortunately, it would seem not all content is available via HLS.  Take https://www.filmon.com/vod/view/277-judex for example, it only returns `.mp4` streams:
```
    "streams": {
      "low": {
        "is_free": true,
        "quality": "low",
        "url": "https://hls-dvr-gv.filmon.com/etv/Judex_1963_Channing_Pollock_Francine_Berge..mp4",
        "name": "SD",
        "watch-timeout": 86000
      },
      "high": {
        "is_free": true,
        "quality": "high",
        "url": "https://hls-dvr-gv.filmon.com/etv/Judex_1963_Channing_Pollock_Francine_Berge..mp4",
        "name": "HD",
        "watch-timeout": 86000
      }
    }
```
whereas others, such as https://www.filmon.com/vod/view/8762-john-lennon-the-ultimate-collection only return HLS streams:
```
    "streams": {
      "low": {
        "is_free": true,
        "quality": "low",
        "url": "https://hls-dvr-gv.filmon.com/uk01.filmon.com/vod/_definst_/content_filmon/source/141/7476/108477/108477/playlist.m3u8",
        "name": "SD",
        "watch-timeout": 86000
      },
      "high": {
        "is_free": true,
        "quality": "high",
        "url": "https://hls-dvr-gv.filmon.com/uk01.filmon.com/vod/_definst_/content_filmon/source/141/7476/108560/108560/playlist.m3u8",
        "name": "HD",
        "watch-timeout": 86000
      }
    }
```
I agree that the warning message is not great and I fully expected you to mention it!

Reply:
After a bit of searching, I found the following project: https://github.com/danielgtaylor/qtfaststart

It appears to have a licence that would allow me to pull the relevant bits out of it and include them within a Streamlink plugin.  I'm very glad somebody else has already done this side of the work; I did not fancy trying to understand this via the specifications, even if I could get the official ones, but ISO/IEC 14496-12:2022 is 198 CHF (186.71 EUR)!!!

Here's the output of before and after using the tool on an mp4 file downloaded from FilmOn (I modified the output to show the position values as well):
```
$ qtfaststart/bin/qtfaststart -l test-orig.mp4 
ftyp at pos: 0 (32 bytes)
free at pos: 32 (8 bytes)
mdat at pos: 40 (276024271 bytes)
moov at pos: 276024311 (4438061 bytes)

$ qtfaststart/bin/qtfaststart -l test-mod.mp4 
ftyp at pos: 0 (32 bytes)
moov at pos: 32 (4438061 bytes)
mdat at pos: 4438093 (276024271 bytes)
```
The modified mp4 file does indeed seem to HTTP stream correctly with a test plugin I created.

I'll have a play with this and see if I can get it all working.

Reply:
Thanks for taking a look at this, but that's way too much of an overkill just for that single plugin. And writing this for a specific plugin is not a good idea and should ideally be done in a more generic way, eg. via a `HTTPStreamMPEG4` class which detects if the moov atom is at the end and adjusts the output automatically, so it can be used by other plugins as well, regardless whether the mp4 file is suitable for streaming via http or not and needs fixing.

Btw, there's a fork with some bugfixes here, since the repo you've linked is unmaintained since a long time ago:
https://github.com/danielyaa5/qtfaststart2

I'd say just leave it as it is. Maybe add a note to the docs instead of outputting a warning message, but I don't really mind having a warning message in the plugin.

Reply:
OK.  Yes, a more generic way is probably the way to go for an implementation.

Thanks for the link to the maintained repo.

I don't have strong feelings either way about the warning message.  If you are happy to keep it and have no other issues with the PR, then you can merge it as is I guess.  Otherwise, I can put a note in the docs and remove the mp4 detection, so the code will just fall through to the "Unsupported stream type" message instead.

Reply:
As said, this can't be merged in the current state because no CI checks have run and passed as there was an outage at GH, and the master branch has certain protection rules when merging. You will have to push again.